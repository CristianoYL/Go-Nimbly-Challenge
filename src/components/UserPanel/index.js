import UserPanel from './user_panel';

export default UserPanel;
