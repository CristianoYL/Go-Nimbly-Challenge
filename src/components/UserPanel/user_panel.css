.user-panel {
  background-color: var(--color-accent);
  color: var(--color-primary);
  display: flex;
  flex-direction: column;
  height: 100%;
}

.text-field-container {
  margin: auto;
  width: 80%;
  display: flex;
  flex-flow: column nowrap;
  align-items: center;
}

.full-width {
  width: 100%;
}

.message-text {
  align-self: flex-start;
  color: var(--color-primary);
}

.btn-submit {
  width: auto;
  margin-top: 20px;
  border: none;
  background-color: transparent;
  color: var(--color-primary);
  align-self: flex-end;
}

.btn-submit:active {
  text-shadow: 1px 1px var(--color-primary);
}

.btn-submit:focus {
  outline: none;
}
