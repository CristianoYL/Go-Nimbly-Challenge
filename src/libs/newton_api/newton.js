/*
 encasulate the newton api caller into a module thus we can make use of it
 more conveniently everywhere in our code.
*/
const askNewton = (operation, expression) => {
  // remove illegal chars
  let cleanExpression = expression.replace(/[^a-z0-9A-Z+\-/*().\^]/g, '');
  // replace / with (over) as per required by API
  cleanExpression = cleanExpression.replace(/[/]/g, '(over)');
  const url = `https://newton.now.sh/${operation}/${cleanExpression}`;
  return fetch(url)
    .then((res) => {
      if (res.ok) {
        return res.json();
      }
      throw new Error('Newton API Server Error.');
    })
    .then((jsonData) => {
      if (jsonData.result.includes('error')) {
        throw new Error('Invalid Syntax found in your input.');
      } else {
        return jsonData;
      }
    })
    .catch((/* e */) => {
      // console.log(e);
      throw new Error('An error has occurred.');
    });
};

export { askNewton };
