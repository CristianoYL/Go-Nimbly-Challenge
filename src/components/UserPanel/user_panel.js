import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './user_panel.css';
import TextField from '../TextField';
import DropdownMenu from '../DropdownMenu';
import { askNewton } from '../../libs/newton_api/newton';

export default class UserPanel extends Component {
  constructor(props) {
    super(props);

    this.state = {
      operation: '',
      expression: '',
    };

    this.getMathResult = this.getMathResult.bind(this);
    this.handleKeyPress = this.handleKeyPress.bind(this);
  }


  getMathResult() {
    if (this.state.operation === '' || this.state.expression === '') {
      return;
    }
    askNewton(this.state.operation, this.state.expression)
      .then((result) => {
        // console.log(jsonData);
        this.props.updateResult(result);
      })
      .catch((e) => {
        // console.log(e);
        this.props.updateResult('Oops, an error has occurred.');
      });
  }

  handleKeyPress(event) {
    console.log(event.key);
    if (event.key === 'Enter') {
      console.log("called");
      this.getMathResult();
    }
  }

  render() {
    return (
      <div className="user-panel">
        <div className="text-field-container">
          <div className="message-text">
            <h2>{this.props.message}</h2>
          </div>

          <DropdownMenu
            className="full-width"
            hint="select a math operation"
            onSelected={operation => this.setState({ operation })}
          >
            <option value="simplify">Simplify</option>
            <option value="factor">Factor</option>
            <option value="derive">Derive</option>
            <option value="integrate">Integrate</option>
            <option value="zeroes">Find 0s</option>
          </DropdownMenu>


          <TextField
            className="full-width"
            hint="your math expression*"
            onTextChange={expression => this.setState({ expression })}
            onKeyPress={(key) => {this.handleKeyPress(key)}}
          />

          <button className="btn-submit" onClick={this.getMathResult}>
            Get Result
          </button>
        </div>
      </div>
    );
  }
}

UserPanel.propTypes = {
  message: PropTypes.string.isRequired,
  updateResult: PropTypes.func.isRequired,
};
