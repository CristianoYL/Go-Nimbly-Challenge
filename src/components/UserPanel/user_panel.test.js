import React from 'react';
import renderer from 'react-test-renderer';
import UserPanel from './user_panel';
import { askNewton } from '../../libs/newton_api/newton';

jest.mock('../../libs/newton_api/newton');

const updateResult = () => {};

test('User panel renders', () => {
  const component = renderer.create(<UserPanel message="test" updateResult={updateResult} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
