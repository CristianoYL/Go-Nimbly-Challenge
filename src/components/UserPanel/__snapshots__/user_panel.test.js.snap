// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User panel renders 1`] = `
<div
  className="user-panel"
>
  <div
    className="text-field-container"
  >
    <div
      className="message-text"
    >
      <h2>
        test
      </h2>
    </div>
    <div
      className="menu-container full-width"
    >
      <label
        htmlFor="operation-select"
        target="_select"
      >
        select a math operation
        <select
          id="operation-select"
          onChange={[Function]}
          value="default"
        >
          <option
            disabled={true}
            value="default"
          >
            -- select an operation --
          </option>
          <option
            value="simplify"
          >
            Simplify
          </option>
          <option
            value="factor"
          >
            Factor
          </option>
          <option
            value="derive"
          >
            Derive
          </option>
          <option
            value="integrate"
          >
            Integrate
          </option>
          <option
            value="zeroes"
          >
            Find 0s
          </option>
        </select>
      </label>
    </div>
    <div
      className="text-field full-width"
    >
      <label
        htmlFor="your math expression*"
        target="_input"
      >
        your math expression*
        <input
          id="your math expression*"
          onChange={[Function]}
          onKeyPress={[Function]}
          placeholder="your math expression*"
          value=""
        />
      </label>
    </div>
    <button
      className="btn-submit"
      onClick={[Function]}
    >
      Get Result
    </button>
  </div>
</div>
`;
